File: backend/app/schemas/tts.py
from typing import Dict, List, Optional

from pydantic import BaseModel, Field


class TTSRequest(BaseModel):
    text: str
    voice_description: Optional[str] = None
    is_narrator: bool = False
    role: Optional[str] = None
    character_id: Optional[str] = None
    binding_override: Optional[Dict] = None


class TTSResponse(BaseModel):
    success: bool
    text: str
    voice_id: str
    gender: str
    provider_id: str
    engine_type: str = "browser"
    is_dialogue: bool
    is_narrator: bool
    role: str = "character"
    clone_sample_id: Optional[str] = None
    speed: float
    volume: float
    style: Optional[str] = None


class TTSConfigRequest(BaseModel):
    enabled: Optional[bool] = None
    active_provider_id: Optional[str] = None
    default_voice_gender: Optional[str] = None
    default_voice_id: Optional[str] = None
    segmented_playback: Optional[bool] = None
    providers: Optional[List[dict]] = None
    provider_configs: Optional[dict] = None


class CustomProviderRequest(BaseModel):
    id: str
    name: str
    description: Optional[str] = ""
    engine_type: str = "custom_api"
    config_fields: Optional[List[dict]] = None
    config: Optional[dict] = None
    voices: Optional[List[dict]] = None


class SetVoiceRequest(BaseModel):
    voice_id: str
    gender: Optional[str] = "female"


class VoiceBindingPayload(BaseModel):
    role: str = "character"
    provider_id: Optional[str] = None
    voice_id: Optional[str] = None
    gender: Optional[str] = None
    clone_sample_id: Optional[str] = None
    speed: float = Field(default=1.0, ge=0.25, le=4.0)
    volume: float = Field(default=1.0, ge=0.0, le=2.0)
    enabled: bool = True
    inherit: bool = False


class BindingsUpdateRequest(BaseModel):
    bindings: List[VoiceBindingPayload]


class TTSPreviewRequest(BaseModel):
    text: Optional[str] = None
    voice_description: Optional[str] = None
    role: str = "character"
    character_id: Optional[str] = None
    binding_override: Optional[Dict] = None


class TTSPrefetchVoicesRequest(BaseModel):
    preview_text: Optional[str] = None


class CloneSampleCreateResponse(BaseModel):
    id: str
    name: str
    provider_id: str
    source_voice_id: Optional[str] = None
    filename: str
    file_size: int
    mime_type: Optional[str] = None

File: frontend/src/types/tts.ts
export type TTSRole = 'character' | 'narrator'

export interface TTSSegment {
  type: 'dialogue' | 'narration'
  text: string
  characterId?: string
  characterName?: string
}

export interface TTSVoiceOption {
  voice_id: string
  gender?: string
  description?: string
  provider_id?: string
  is_clone?: boolean
  clone_sample_id?: string
}

export interface TTSProvider {
  id: string
  name: string
  description?: string
  engine_type: 'browser' | 'xiaomi_mimo' | 'custom_api' | string
  voices?: TTSVoiceOption[]
  is_builtin?: boolean
  config_fields?: Array<Record<string, unknown>>
  config?: Record<string, string>
}

export interface TTSVoiceBinding {
  id?: string
  scope?: 'global' | 'user' | 'character'
  user_id?: number | null
  character_id?: string | null
  role: TTSRole
  provider_id?: string | null
  voice_id?: string | null
  gender?: string | null
  clone_sample_id?: string | null
  speed?: number
  volume?: number
  enabled?: boolean
  inherited?: boolean
}

export interface TTSResolvedVoice {
  provider_id: string
  engine_type?: string
  voice_id: string
  gender: string
  clone_sample_id?: string | null
  role: TTSRole
  text?: string
  speed?: number
  volume?: number
  source?: string
}

export interface TTSBindingState {
  explicit?: TTSVoiceBinding | null
  resolved: TTSResolvedVoice
}

export interface TTSCloneSample {
  id: string
  name: string
  provider_id: string
  source_voice_id?: string | null
  filename: string
  file_size: number
  mime_type?: string | null
  duration_seconds?: number | null
  created_at?: string
  updated_at?: string
  usage_count?: number
}

export interface TTSManagementState {
  enabled: boolean
  active_provider_id: string
  segmented_playback: boolean
  providers: TTSProvider[]
  voices: TTSVoiceOption[]
  global_bindings: Record<TTSRole, TTSVoiceBinding | null>
  my_bindings: Record<TTSRole, TTSBindingState>
  clone_samples: TTSCloneSample[]
  can_admin: boolean
}

export interface TTSBindingPayload {
  role: TTSRole
  provider_id?: string | null
  voice_id?: string | null
  gender?: string | null
  clone_sample_id?: string | null
  speed?: number
  volume?: number
  enabled?: boolean
  inherit?: boolean
}

export interface TTSPreviewRequest {
  text?: string
  voice_description?: string
  role: TTSRole
  character_id?: string
  binding_override?: Partial<TTSBindingPayload>
}

export interface TTSPreviewMetadata extends TTSResolvedVoice {
  success: boolean
  is_dialogue: boolean
  is_narrator: boolean
}
